Persist storage API

ginge opened the pull request:
implement pebble persist_x apis on flash

jwise commented on rcore/rdb.c:
```diff
@@ -517,6 +529,49 @@ int rdb_insert(const struct rdb_database *db, uint8_t *key, uint16_t key_size, u
+    rdb_select_result_foreach(res, &head) {
+        // is the data size <=
+        if (data_size != res->it.data_len) {
+            if (fs_write(&res->it.fd, data, data_size) < data_size) {
```
This only works if the data is a subset (i.e., `old_data & new_data == new_data`) of the old data, right?

My plan in general was a generalized update mechanism on top of `insert` that always did a replace on an old key (I think there is a way to safely do it atomically), and then a `rdb_amend` op that works like this branch does, and demands that the old size be equal to the new size.

ginge replied:
that sounds good. One thing I noticed was an app continually writing to the same storage rdb. The data size was always the same, so I figured we just update the value save thrashing flash.
If the data size is not the same, we do the delete insert dance.
(also just noticed the comment doesn't match the logic any more)
I welcome our atomic update overlord

ginge replied:
It has just occured to me what you mean by this. Updating the flash in place isn't gonna work is it

jwise commented on rwatch/storage/storage_persist.c:
```diff
@@ -0,0 +1,243 @@
+    rdb_select_result_list head;
+    list_init_head(&head);
+    
+    struct rdb_database *db = rdb_open(RDB_ID_APP_PERSIST);
```
I wonder if we need a mechanism to isolate apps from each other (i.e., to prevent an app from filling up the database) -- that is to say, to give each app its own rdb.

ginge replied:
yes, I agree. The composite key was a quick way to achieve this without epic plumbing

ginge replied:
(also would be easier at ensuring per-app storage limits)

jwise commented on rwatch/storage/storage_persist.c:
```diff
@@ -0,0 +1,243 @@
+    struct App *app = appmanager_get_current_app();
+    assert(app);
+    
+    app_persist_key c_key = { .appid = app->id, .key = key };    
```
Not blocking this merge, but a TODO in the future: is app->id the correct thing here, or should it be app->uuid?  Otherwise when we upload a new version of the app (or the app gets uninstalled and reinstalled) it'd get a new id, and you'd lose the db, I think?

ginge replied:
yes, actually you make a really good point here. I'll probably look at changing this before we merge

jwise commented:
Shiny as heck.

ginge commented:
I've addressed the flash update in place mechanism. You're absolutely right on the NAND a & b being the only use case that would work.
switched to UUID for apps. Makes more sense
